dotnet/src/Connectors/Connectors.AI.Oobabooga/TextCompletion/OobaboogaTextCompletion.cs

```diff
@@ -0,0 +1,257 @@
+    private const string ResponseObjectTextStreamEvent = "text_stream";
+    private const string ResponseObjectStreamEndEvent = "stream_end";
+
+    private readonly Uri _endpoint;
+    private readonly int _blockingPort;
+    private readonly int _streamingPort;
+    private readonly HttpClient _httpClient;
+    private readonly ClientWebSocket? _webSocket;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="OobaboogaTextCompletion"/> class.
+    /// </summary>
+    /// <param name="endpoint">Endpoint for service API call.</param>
+    /// <param name="blockingPort">The port for blocking requests.</param>
+    /// <param name="streamingPort">The port for streaming requests.</param>
+    /// <param name="httpClient">The HTTP client to use for making regular blocking API requests. If not specified, a default client will be used.</param>
+    /// <param name="webSocket">The client web socket to use for making streaming API requests. If not specified, a transient client will be used on each call.</param>
+    public OobaboogaTextCompletion(Uri endpoint, int blockingPort, int streamingPort, HttpClient? httpClient = null, ClientWebSocket? webSocket = null)
+    {
+        Verify.NotNull(endpoint);
+
+        this._endpoint = endpoint;
+        this._blockingPort = blockingPort;
+        this._streamingPort = streamingPort;
+        this._httpClient = httpClient ?? new HttpClient(NonDisposableHttpClientHandler.Instance, disposeHandler: false);
+        if (webSocket != null)
+        {
+            this.SetWebSocketOptions(webSocket);
+            this._webSocket = webSocket;
+        }
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<ITextStreamingResult> GetStreamingCompletionsAsync(
+        string text,
+        CompleteRequestSettings requestSettings,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        UriBuilder streamingUri = new(this._endpoint)
+        {
+            Port = this._streamingPort,
+            Path = StreamingUriPath
+        };
+        if (streamingUri.Uri.Scheme.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+        {
+            streamingUri.Scheme = (streamingUri.Scheme == "https") ? "wss" : "ws";
+        }
+
+        var completionRequest = this.CreateOobaboogaRequest(text, requestSettings);
+
+        var requestJson = JsonSerializer.Serialize(completionRequest);
+
+        var requestBytes = Encoding.UTF8.GetBytes(requestJson);
+        ClientWebSocket? transientClientWebSocket = null;
+        try
+        {
+            ClientWebSocket? clientWebSocket;
+            if (this._webSocket is null)
+            {
+                transientClientWebSocket = new();
+                this.SetWebSocketOptions(transientClientWebSocket);
+                clientWebSocket = transientClientWebSocket;
+            }
+            else
+            {
+                clientWebSocket = this._webSocket;
```

Comment:
What's going to happen if the GetStreamingCompletionsAsync method is called more than once in parallel and the first call already sent a prompt to the server and is waiting/reading completion/response from it while the second call sends a new prompt to the same server through the same instance of ClientWebSocket - _webSocket ?

Reply:
OK so let's summarize the situation.
Websocket clients have those distinctive features:

- They cannot be replaced using a HttpClient since they use a distinct protocol. An HttpListener can be used on the server side though as in the mock server, but that's about it.
- Unlike HttpClient, they don't have a non disposable version, and their handler isn't available, at least until incoming versions.
- As you noted, they're not meant for concurrent access.

The initial version of the connector in this PR had a single websocket client properly disposed.
Then the option was introduced to passe the websocket client as part of the constructor.
Then it was requested that the connector be non disposable, and since there is no non-disposable version of websocket client I introduced those transient version, while keeping the option to provide one externally in the ctor.

Now both your remark make sense, but they put us in a tricky situation: We should avoid single use websocket clients to prevent port exhaustion in the extreme case, but we should also avoir reusing websocket clients in concurrent scenarios to avoid unintended behavior, added to the constraint that we must support both clients supplied externally and instantiated internally, and that we cannot dispose the latter with the class.

That was quite a challenge but I tried to respect those 4 hard constraints, with some kind of convoluted pooling: Incoming client calls signal concurrency by pushing and poping a flag to a concurrent stack. When the stack gets empty means there is no concurrent call anymore. As long as the stack is not empty, used clients get recycled in another concurrent stack, and when there is no concurrency, last call flushes and disposes the existing websocket clients before leaving.
As for externally provided websocket clients, I replaced the original ctor paramter with a factory, leaving the responsibility to deal with pooling to the consumer, with the corresponding comment warning about hazardous concurrent use and port exhaustion.

I also fixed the mock websocket server and added new unit tests, to check websocket server concurrency, and to test concurrent streaming calls with multiple websocket response messages, both in the persistent and the transient cases.

Hopefully this will be enough to merge that PR, since as long as it stays there, I have to deal with the numerous changes to the exposed surface, which has been exhausting. Also, my previous comments addressed a bug, which I spent hours chasing. Can you acknowledge that my fix is sound?

Reply:
@SergeyMenshykh Please check [this test](https://github.com/microsoft/semantic-kernel/pull/1357/files#diff-a76773b8dde48313e8b84fcfeac12a63334d1ceea3e5d7010882c2a108373973R243) and [that one](https://github.com/microsoft/semantic-kernel/pull/1357/files#diff-a76773b8dde48313e8b84fcfeac12a63334d1ceea3e5d7010882c2a108373973R257) where a lot of concurrent calls are made. Distinct websockets are used for each request, but they get recycled to prevent resource exhaustion as per your previous comment.